experimental/Geometry/Schemes.jl

```diff
@@ -0,0 +1,1244 @@
+# ring of a polynomial algebra.
+mutable struct Spec{S,T,U} <: AffineScheme{S,T,U}
+  # the basic fields 
+  k::S			# the base ring (usually a field) of definition for the scheme
+  R::T		  	# the ambient polynomial ring to model this affine scheme
+  I::MPolyIdeal{U}	# The ideal in R defining the scheme
+
+  # fields for caching
+  name::String # the name of this scheme for printing
+
+  function Spec(k::S, R::T, I::MPolyIdeal{U} ) where{
+			S <: Ring, T <:MPolyRing , U <: MPolyElem}
+    if k != coefficient_ring(R)
+      error( "Base ring of the affine scheme does not coincide with the base ring of the associated algebra." )
+    end
+    # TODO: Implement further plausibility checks to be performed at runtime.
+    return new{S, T, U}(k, R, I )
+  end
+end
+
+###################################################################################
+# Getter functions
+#
+# No caching is needed in this case, since all these variables need to be assigned 
+# at instantiation
+#
+function base_ring(A::Spec)
+  return A.k
+end
+
+function ambient_ring(A::Spec)
+  return A.R
+end
+
+function defining_ideal(A::Spec)
+  return A.I
+end
+
+##################################################################################
+# Affine scheme that arises as the localization of a Spec at a specific 
+# element 'denom' of the coordinate ring, i.e. a principal open subset. 
+#
+# These are implemented in a recursive fashion meaning there is a parent 
+# which is either a classical Spec or another principal open subset, from 
+# which this instance is derived by inverting an element 'denom'. 
+#
+# Such parent structures form a tree which necessarily has a Spec at its 
+# root. Currently, the implementation only allows localizations at elements 
+# denom from the coordinate ring of this root. 
+#
+mutable struct SpecPrincipalOpen{S,T,U} <: AffineScheme{S,T,U}
+  parent::AffineScheme{S,T,U}
+  denom::U  # element of the "ambient" polynomial ring of the root
+  	    # which has been localized in the parent to get this scheme.
+
+  # Fields for caching. These provide the data as a Spec for this 
+  # affine scheme
+  k::S  # the ground field
+  R::T	# the ambient polynomial ring
+  I::MPolyIdeal{U} # the defining ideal
+  pullbackFromParent::AlgHom
+  pullbackFromRoot::AlgHom
+  u::U # The inverse of denom in R/I (one of the variables)
+  name::String
+  #inclusionInParent::AffSchMorphism
+  #inclusionInRoot::AffSchMorphism
+
+  
+  function SpecPrincipalOpen(parent::Union{Spec{S,T,U},SpecPrincipalOpen{S,T,U}}, denom::U) where {S <: Ring, T<:MPolyRing, U<:MPolyElem}
+    x = new{S,T,U}() 
+    x.parent = parent
+    # TODO: Implement a plausibility check: Does denom belong to the coordinate ring of the root?
+    x.denom = denom
+    return x 
+  end 
+end
+
+###############################################################################
+# Getter functions
+
+function PrincipalSubScheme(parent, denom)
```

Comment:
Only types (and by necessity, constructors) should use `CamelCase`; functions should use `snake_case`.

Since this distinction may be unclear and fuzzy for end users, and also sometimes developers, we also tend to have alternate constructors using `snake_case`; I can't judge whether this would make sense here (i.e. having a `spec_principal_open` function) but what I can say is that `PrincipalSubScheme` isn't a type, so it shouldn't have this name.

Reply:
> The tests should be moved correspondingly to `test/experimental/Geometry/Schemes.jl `

I don't find any such folder: 'test/experimental'. Are you sure?

Reply:
As far as I know the directory structure of the test files should mirror that of the source code, so that we know where to look for the tests. If that folder is not there, you can create it.

Reply:
We haven't been very consistent about this, to be fair. E.g. the plane curves code by @delphinepol is now in `experimental` but the tests are still in `test/Examples/PlaneCurve-test.jl`.
 
So I think this is not *that* important right now...

But what I *do* wonder is whether perhaps `Geometry` should be changed to `AlgebraicGeometry`? We already have "Polyhedral Geometry" (and who knows, one day we might also have Differential/Incidence/Arithmetic/... geometry?).

Reply:
My point about the folder was that not only "Geometry" doesn't exist there, but really the whole test folder for the experimental section can't be found. It seems that nobody who ever contributed to the experimental section put their tests anywhere else but in the usual test folder.

Reply:
About the name: I have no strong feelings about the change there. AlgebraicGeometry would be fine with me.